### PR TITLE
Workaround sympy modulus bugginess

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113452
* __->__ #113878
* #113877

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Comment by ezyang** [comment]
Actually, sympy.Mod doesn't have this problem but our Mod does, so I should probably fix our Mod instead.

**Comment by ezyang** [comment]
I put a breakpoint in our Mod.eval implementation and it correctly returns None saying no simplification is possible. So why does sympy.Eq(Mod(s0, 2.0), 0.0) return False.....

**Comment by asmeurer** [comment]
What are the assumptions that get set on `s0`? 

At any rate, I suspect this was fixed with https://github.com/sympy/sympy/pull/25875

**Comment by ezyang** [comment]
OK, maybe I will keep this workaround for now and then remove it when the sympy fix makes it public

**Comment by ezyang** [comment]
The assumptions are that s0 is integer, is positive

**Comment by asmeurer** [comment]
Just tested SymPy master against this branch and it does indeed fix the issue.

File: torch/fx/experimental/sym_node.py [hunk 1]
```diff
@@ -793,6 +804,14 @@ def binary_magic_impl(self, other):
                 self, handle_sym_dispatch(op, (wrap_node(self), wrap_node(other)), {})
             )
         assert isinstance(other, SymNode)
+
+        # Sympy will incorrectly conclude that Mod(s0, 2.0) can never equal
+        # 0.0.  So if we attempt to do a modulus with float, just force
+        # specialization and stop trying to symbolically reason about it
+        if method == "mod" and (self.pytype is float or other.pytype is float):
+            out = self.guard("", 0) % other.guard("", 0)
```
Comment (bdhirsh) [inline torch/fx/experimental/sym_node.py:812]: Why the empty file / line_no, instead of using the the file/line_no where the mod call came from?

**Comment by ezyang** [comment]
I'm abandoning this PR, the cure is worse than the disease. Best to fix it in sympy.

**Pull request closed without merging**